src/util/expr_iterator.h

```diff
@@ -215,17 +227,51 @@ class const_depth_iteratort final:
 class depth_iteratort final:
   public depth_iterator_baset<depth_iteratort>
 {
+private:
+  std::function<exprt &()> mutate_root;
+
 public:
   /// Create an end iterator
   depth_iteratort()=default;
   /// Create iterator starting at the supplied node (root)
   /// All mutations of the child nodes will be reflected on this node
-  explicit depth_iteratort(exprt &expr):
-    depth_iterator_baset(expr) { }
+  explicit depth_iteratort(exprt &expr) : depth_iterator_baset(expr)
+  {
+  }
+  /// Create iterator starting at the supplied node (root)
+  /// All mutations of the child nodes will be reflected on this node
+  explicit depth_iteratort(
+    const exprt &expr,
+    std::function<exprt &()> mutate_root)
+    : depth_iterator_baset(expr), mutate_root(mutate_root)
+  {
+    PRECONDITION(mutate_root);
```

Comment:
Why is this precondition relevant? Wouldn't a `nullptr` just be the same as if invoking the other constructor?

Reply:
The other constructor requires a non-const exprt. I'll add a comment to this line explaining.